src/pipeline/pipeline_construction.c

```diff
@@ -200,26 +239,55 @@ static ResultProcessor *getArrangeRP(Pipeline *pipeline, const AggregationPipeli
         ResultProcessor *rpLoader = RPLoader_New(params->common.sctx, params->common.reqflags, lk, loadKeys, array_len(loadKeys), forceLoad, outStateFlags);
         up = pushRP(&pipeline->qctx, rpLoader, up);
       }
-      rp = RPSorter_NewByFields(maxResults, sortkeys, nkeys, astp->sortAscMap);
-      up = pushRP(&pipeline->qctx, rp, up);
+
+      if (IsAggregate(&params->common) && HasWithCount(&params->common) && HasSortBy(&params->common)) {
+        size_t sort_limit = getSortLimit(params, astp, maxResults);
+        rp = RPSorter_NewByFields(sort_limit, sortkeys, nkeys, astp->sortAscMap);
+        up = pushRP(&pipeline->qctx, rp, up);
+      } else {
+        rp = RPSorter_NewByFields(maxResults, sortkeys, nkeys, astp->sortAscMap);
+        up = pushRP(&pipeline->qctx, rp, up);
+      }
     } else if (IsHybrid(&params->common) ||
```

Comment:
Consider simplifying?
```suggestion
      size_t sort_limit = maxResults;
      if (IsAggregate(&params->common) && HasWithCount(&params->common) && HasSortBy(&params->common)) {
        sort_limit = getSortLimit(params, astp, maxResults);
      }
      rp = RPSorter_NewByFields(sort_limit, sortkeys, nkeys, astp->sortAscMap);
        up = pushRP(&pipeline->qctx, rp, up);
    } else if (IsHybrid(&params->common) ||
```